\documentclass{article}
\usepackage[utf8]{inputenc}
\usepackage{hyperref}
\usepackage{graphicx}
\usepackage{listings}
\usepackage{amssymb}
\usepackage{color}
\usepackage{xcolor}
\usepackage{indentfirst}
\usepackage{makeidx}

\makeindex


\lstset{language=bash,frame=single,backgroundcolor=\color{lightgray},numbers=left,breaklines=true,breakautoindent=true,basicstyle=\small}
\date{\today}
\title{GNU Parallel for Bioinformatics}
\author{Pierre Lindenbaum\\\href{https://twitter.com/yokofakun}{@yokofakun}\\\url{http://plindenbaum.blogspot.com} }


\def\prl{\textbf{parallel}}
\def\bam{\textbf{bam}}
\def\BAM{\textbf{BAM}}
\def\samtools{\textbf{samtools}}

\begin{document}
\maketitle
\begin{abstract}
This document follows the Ole Tange's \prl{} tutorial \url{http://www.gnu.org/software/parallel/parallel_tutorial.html}.
\end{abstract}

\tableofcontents

\section{Input Source}
\subsection{A single input source}
\subsubsection{Input can be read from the command line.}
\begin{lstlisting}
$ parallel file ::: samtools-0.1.18/examples/*.bam
\end{lstlisting}
output:
\begin{lstlisting}
samtools-0.1.18/examples/ex1a.bam: gzip compressed data, extra field
samtools-0.1.18/examples/ex1.bam: gzip compressed data, extra field
samtools-0.1.18/examples/ex1b.bam: gzip compressed data, extra field
samtools-0.1.18/examples/ex1f.bam: gzip compressed data, extra field
samtools-0.1.18/examples/ex1f-rmduppe.bam: gzip compressed data, extra field
samtools-0.1.18/examples/ex1f-rmdupse.bam: gzip compressed data, extra field
samtools-0.1.18/examples/ex1_sorted.bam: gzip compressed data, extra field
samtools-0.1.18/examples/toy.bam: gzip compressed data, extra field
\end{lstlisting}

\subsubsection{The input source can be a file}
\begin{lstlisting}
$ find samtools-0.1.18/examples/ -name "*.bam" -type f > listbams.txt
$ parallel -a listbams.txt file
\end{lstlisting}
output:
\begin{lstlisting}
samtools-0.1.18/examples/ex1a.bam: gzip compressed data, extra field
samtools-0.1.18/examples/ex1.bam: gzip compressed data, extra field
samtools-0.1.18/examples/ex1b.bam: gzip compressed data, extra field
samtools-0.1.18/examples/ex1f.bam: gzip compressed data, extra field
samtools-0.1.18/examples/ex1f-rmduppe.bam: gzip compressed data, extra field
samtools-0.1.18/examples/ex1f-rmdupse.bam: gzip compressed data, extra field
samtools-0.1.18/examples/ex1_sorted.bam: gzip compressed data, extra field
samtools-0.1.18/examples/toy.bam: gzip compressed data, extra field
\end{lstlisting}




\subsubsection{STDIN (standard input) can be the input source}
\begin{lstlisting}
$ find samtools-0.1.18/examples/ -name "*.bam" -type f | parallel file
\end{lstlisting}
output:
\begin{lstlisting}
samtools-0.1.18/examples/ex1a.bam: gzip compressed data, extra field
samtools-0.1.18/examples/ex1.bam: gzip compressed data, extra field
samtools-0.1.18/examples/ex1b.bam: gzip compressed data, extra field
samtools-0.1.18/examples/ex1f.bam: gzip compressed data, extra field
samtools-0.1.18/examples/ex1f-rmduppe.bam: gzip compressed data, extra field
samtools-0.1.18/examples/ex1f-rmdupse.bam: gzip compressed data, extra field
samtools-0.1.18/examples/ex1_sorted.bam: gzip compressed data, extra field
samtools-0.1.18/examples/toy.bam: gzip compressed data, extra field
\end{lstlisting}


Another example: indexing sorted sorted \bam{} files with \samtools{}:
\begin{lstlisting}
$ find dir1 -name "*.bam" | grep sorted |\
	parallel -a -  'samtools index '
\end{lstlisting}
or , without '-a -'
\begin{lstlisting}
$ find dir1 -name "*.bam" | grep sorted |\
	parallel   'samtools index '
\end{lstlisting}

\subsection{Multiple input source}

%%
%% $ for L in 01 02 03 04 05 06 07 08 09 ; do curl -s "http://hgdownload.cse.ucsc.edu/goldenPath/hg19/database/kgXref.txt.gz" | gunzip -c | cut -d '   ' -f 4 | grep _ | uniq | head -n 10 | shuf | head -n 5 | sort > list_genes_${L}.txt  ; done
%%

\begin{lstlisting}
$  parallel echo ::: A T G C ::: A T G C
\end{lstlisting}
output:
\begin{lstlisting}
A A
A T
A G
A C
T A
T T
T G
T C
G A
G T
G G
G C
C A
C T
C G
C C
\end{lstlisting}

\subsubsection{If one of the input sources is too short, its values will wrap}
\begin{lstlisting}
$  parallel echo ::: A T G C N ::: A T G C
\end{lstlisting}

output:
\begin{lstlisting}
A A
A T
A G
A C
T A
T T
T G
T C
G A
G T
G G
G C
C A
C T
C G
C C
N A
N T
N G
N C
\end{lstlisting}

\subsubsection{The input sources can be files}
\begin{lstlisting}
$  parallel  -a <(echo -e "A\nT\nG\nC") -a   <(echo -e "a\nt\ng\nc") echo
\end{lstlisting}
output:
\begin{lstlisting}
A a
A t
A g
A c
T a
T t
T g
T c
G a
G t
G g
G c
C a
C t
C g
C c
\end{lstlisting}


\subsubsection{STDIN can be one of the input sources using '-'}
\begin{lstlisting}
$  echo -e "A\nT\nG\nC" |\
	parallel  -a - -a   <(echo -e "a\nt\ng\nc") echo
\end{lstlisting}
output:
\begin{lstlisting}
A a
A t
A g
A c
T a
T t
T g
T c
G a
G t
G g
G c
C a
C t
C g
C c
\end{lstlisting}

\subsubsection{Instead of -a files can be given after '\texttt{::::}'}
\begin{lstlisting}
parallel  echo :::: <(echo -e "A\nT\nG\nC") ::::   <(echo -e "a\nt\ng\nc")
\end{lstlisting}
output:
\begin{lstlisting}
A a
A t
A g
A c
T a
T t
T g
T c
G a
G g
G t
G c
C a
C t
C g
C c
\end{lstlisting}

\subsubsection{'\texttt{:::}' and '\texttt{::::}' can be mixed:}
\begin{lstlisting}
$ parallel 'grep -Hn ' ::: B7ZGX9 I7FC33 :::: <(ls list_genes_*.txt)
\end{lstlisting}
output:
\begin{lstlisting}
list_genes_01.txt:1:B7ZGX9_HUMAN
list_genes_02.txt:1:B7ZGX9_HUMAN
list_genes_04.txt:1:B7ZGX9_HUMAN
list_genes_07.txt:1:B7ZGX9_HUMAN
list_genes_08.txt:1:B7ZGX9_HUMAN
list_genes_09.txt:1:B7ZGX9_HUMAN
list_genes_02.txt:3:I7FC33_HUMAN
list_genes_06.txt:3:I7FC33_HUMAN
list_genes_07.txt:4:I7FC33_HUMAN
list_genes_08.txt:3:I7FC33_HUMAN
\end{lstlisting}

\subsection{Matching arguments from all input sources}
\subsubsection{With \texttt{--xapply} you can get one argument from each input source}

with \texttt{--xapply}
\begin{lstlisting}
parallel --xapply echo ::: A T G C ::: a t g c n
\end{lstlisting}
output:
\begin{lstlisting}
A a
T t
G g
C c
A n
\end{lstlisting}

without \texttt{--xapply}:
\begin{lstlisting}
$ parallel echo ::: A T G C ::: a t g c n
\end{lstlisting}
output:
\begin{lstlisting}
A a
A t
A g
A c
A n
T a
T t
T g
T c
T n
G a
G t
G g
G c
G n
C a
C t
C g
C c
C n
\end{lstlisting}

\subsubsection{If one of the input sources is too short, its values will wrap}
\begin{lstlisting}
$ parallel --xapply echo ::: A T  ::: g c n
\end{lstlisting}
output:
\begin{lstlisting}
A g
T c
A n
\end{lstlisting}

\subsection{Changing the argument separator}
GNU Parallel can use other separators than ::: or ::::. This is typically useful if ::: or :::: is used in the command to run.
\begin{lstlisting}
$parallel --arg-sep yoyo echo yoyo B7ZGX9 I7FC33 yoyo EIF4G1 PABPC1 yoyo B7ZGX9_HUMAN  C9J4L2_HUMAN
\end{lstlisting}
output:
\begin{lstlisting}
B7ZGX9 EIF4G1 B7ZGX9_HUMAN
B7ZGX9 EIF4G1 C9J4L2_HUMAN
B7ZGX9 PABPC1 B7ZGX9_HUMAN
B7ZGX9 PABPC1 C9J4L2_HUMAN
I7FC33 EIF4G1 B7ZGX9_HUMAN
I7FC33 EIF4G1 C9J4L2_HUMAN
I7FC33 PABPC1 B7ZGX9_HUMAN
I7FC33 PABPC1 C9J4L2_HUMAN
\end{lstlisting}
\subsubsection{Changing the argument file separator}
\begin{lstlisting}
$ ls list_genes_0* | parallel --arg-file-sep schtroumph 'grep -nH' ::: B7ZGX9 I7FC33 schtroumph -
\end{lstlisting}
output:
\begin{lstlisting}
list_genes_01.txt:1:B7ZGX9_HUMAN
list_genes_02.txt:1:B7ZGX9_HUMAN
list_genes_04.txt:1:B7ZGX9_HUMAN
list_genes_07.txt:1:B7ZGX9_HUMAN
list_genes_08.txt:1:B7ZGX9_HUMAN
list_genes_09.txt:1:B7ZGX9_HUMAN
list_genes_02.txt:3:I7FC33_HUMAN
list_genes_06.txt:3:I7FC33_HUMAN
list_genes_07.txt:4:I7FC33_HUMAN
list_genes_08.txt:3:I7FC33_HUMAN
\end{lstlisting}
\subsubsection{Changing the argument delimiter}
\begin{lstlisting}
$ $ echo -n "7ZGX9,I7FC33" | \
	parallel -d , 'grep -nH' :::: - :::  list_genes_0*
\end{lstlisting}
output:
\begin{lstlisting}
list_genes_01.txt:1:B7ZGX9_HUMAN
list_genes_02.txt:1:B7ZGX9_HUMAN
list_genes_04.txt:1:B7ZGX9_HUMAN
list_genes_07.txt:1:B7ZGX9_HUMAN
list_genes_08.txt:1:B7ZGX9_HUMAN
list_genes_09.txt:1:B7ZGX9_HUMAN
list_genes_02.txt:3:I7FC33_HUMAN
list_genes_06.txt:3:I7FC33_HUMAN
list_genes_07.txt:4:I7FC33_HUMAN
list_genes_08.txt:3:I7FC33_HUMAN
\end{lstlisting}

\subsubsection{NULL can be given as  \texttt{\textbackslash{}0}}
\begin{lstlisting}
$  echo -n -e "7ZGX9\0I7FC33" |  parallel -d '\0' 'grep -nH' :::: - :::  list_genes_0*
\end{lstlisting}
output:
\begin{lstlisting}
list_genes_01.txt:1:B7ZGX9_HUMAN
list_genes_02.txt:1:B7ZGX9_HUMAN
list_genes_04.txt:1:B7ZGX9_HUMAN
list_genes_07.txt:1:B7ZGX9_HUMAN
list_genes_08.txt:1:B7ZGX9_HUMAN
list_genes_09.txt:1:B7ZGX9_HUMAN
list_genes_02.txt:3:I7FC33_HUMAN
list_genes_06.txt:3:I7FC33_HUMAN
list_genes_07.txt:4:I7FC33_HUMAN
list_genes_08.txt:3:I7FC33_HUMAN
\end{lstlisting}

\subsubsection{A shorthand for '\texttt{-d  \textbackslash{}0}' is \texttt{-0}}
This will often be used to read files from find ... -print0
\begin{lstlisting}
$ find ./ -name "list_genes_0*.txt" -print0 |\
	parallel -0 'grep -nH'  ::: B7ZGX9 I7FC33 :::: -
\end{lstlisting}
output:
\begin{lstlisting}
./list_genes_01.txt:1:B7ZGX9_HUMAN
./list_genes_04.txt:1:B7ZGX9_HUMAN
./list_genes_02.txt:1:B7ZGX9_HUMAN
./list_genes_09.txt:1:B7ZGX9_HUMAN
./list_genes_07.txt:1:B7ZGX9_HUMAN
./list_genes_08.txt:1:B7ZGX9_HUMAN
./list_genes_06.txt:3:I7FC33_HUMAN
./list_genes_02.txt:3:I7FC33_HUMAN
./list_genes_07.txt:4:I7FC33_HUMAN
./list_genes_08.txt:3:I7FC33_HUMAN
\end{lstlisting}


\section{Building the command line}

\subsubsection{No command means arguments are commands}

If no command is given after parallel the arguments themselves are treated as commands:
\begin{lstlisting}
parallel ::: 'ls -la  toy*' 'file toy.bam' pwd
\end{lstlisting}
output:
\begin{lstlisting}
toy.bam: gzip compressed data, extra field
/path/to/samtools-0.1.18/examples
-rw-rw-r-- 1 lindenb lindenb 478 Mar 27  2013 toy.bam
-rw-rw-r-- 1 lindenb lindenb 176 Mar 27  2013 toy.bam.bai
-rw-rw-r-- 1 lindenb lindenb 254 Mar 27  2013 toy.dict
-rw-r--r-- 1 lindenb lindenb  98 Mar 27  2013 toy.fa
-rw-rw-r-- 1 lindenb lindenb  32 Mar 27  2013 toy.fa.fai
-rw-r--r-- 1 lindenb lindenb 786 Apr 22  2011 toy.sam
-rw-rw-r-- 1 lindenb lindenb 176 Jul 19 09:28 toy_sorted.bai
-rw-rw-r-- 1 lindenb lindenb 478 Aug  6 15:11 toy_sorted.bam
-rw-rw-r-- 1 lindenb lindenb 176 Oct  2 10:24 toy_sorted.bam.bai
\end{lstlisting}

The command can be a script, a binary or a Bash function if the function is exported using 'export -f':
\begin{lstlisting}
$ index_bam_with_samtools() {
echo "Indexing $1" && samtools index $1
}
$ export -f index_bam_with_samtools
$ find ./ -name "*_sorted.bam" |\
   parallel -a - index_bam_with_samtools
\end{lstlisting}
output:
\begin{lstlisting}
Indexing ./ex1b_sorted_sorted.bam
Indexing ./ex1b_sorted.bam
Indexing ./sorted_sorted_sorted.bam
Indexing ./ex1f-rmduppe_sorted.bam
Indexing ./ex1f-rmdupse_sorted.bam
Indexing ./toy_sorted.bam
Indexing ./ex1_sorted_sorted.bam
Indexing ./ex1a_sorted.bam
Indexing ./ex1_sorted.bam
Indexing ./ex1f-rmduppe_sorted_sorted.bam
Indexing ./ex1f-rmdupse_sorted_sorted.bam
Indexing ./ex1f_sorted_sorted.bam
Indexing ./ex1a_sorted_sorted.bam
Indexing ./sorted_sorted.bam
Indexing ./ex1f_sorted.bam
\end{lstlisting}

\section{Replacement strings}
\subsection{The 5 replacement strings}
GNU Parallel has several replacement strings. If no replacement strings are used the default is to append {}:
\begin{lstlisting}
$ parallel  'grep -Hn ">" ' ::: toy.fa ex1.fa
toy.fa:1:>ref
toy.fa:3:>ref2
ex1.fa:1:>seq1
ex1.fa:29:>seq2
\end{lstlisting}
output:
\begin{lstlisting}
toy.fa:1:>ref
toy.fa:3:>ref2
ex1.fa:1:>seq1
ex1.fa:29:>seq2
\end{lstlisting}

The default replacement string is \texttt{'\{\}'}:
\begin{lstlisting}
$ parallel  'grep -Hn ">" ' {} ::: toy.fa ex1.fa
\end{lstlisting}
output:
\begin{lstlisting}
toy.fa:1:>ref
toy.fa:3:>ref2
ex1.fa:1:>seq1
ex1.fa:29:>seq2
\end{lstlisting}

The replacement string \texttt{'\{\}'} can be changed with \texttt{'-I'}:
\begin{lstlisting}
$ parallel -I FILE_NAME 'grep -Hn ">" ' FILE_NAME ::: toy.fa ex1.fa
\end{lstlisting}
output:
\begin{lstlisting}
toy.fa:1:>ref
toy.fa:3:>ref2
ex1.fa:1:>seq1
ex1.fa:29:>seq2
\end{lstlisting}

The replacement string  \texttt{'\{.\}'} removes the extension:

Sorting the BAMs:
\begin{lstlisting}
$ find ./ -name "*.bam" |\
  parallel -a - 'samtools sort' {} {.}_sorted &&\
  find ./ -name "*_sorted.bam"
\end{lstlisting}
output:
\begin{lstlisting}
./examples/ex1b_sorted.bam
./examples/ex1f-rmduppe_sorted.bam
./examples/ex1f-rmdupse_sorted.bam
./examples/toy_sorted.bam
./examples/ex1a_sorted.bam
./examples/ex1_sorted.bam
./examples/ex1f_sorted.bam
\end{lstlisting}

The replacement string \texttt{'\{.\}'} can be changed with  \texttt{'--extensionreplace'}:
Sorting the BAMs:
\begin{lstlisting}
$ find ./ -name "*.bam" |\
  parallel --extensionreplace BARBAPAPA -a - 'samtools sort' {} BARBAPAPA_sorted &&\
  find ./ -name "*_sorted.bam"
\end{lstlisting}
output:
\begin{lstlisting}
./examples/ex1b_sorted.bam
./examples/ex1f-rmduppe_sorted.bam
./examples/ex1f-rmdupse_sorted.bam
./examples/toy_sorted.bam
./examples/ex1a_sorted.bam
./examples/ex1_sorted.bam
./examples/ex1f_sorted.bam
\end{lstlisting}



The replacement string \texttt{'\{/\}'}: removes the path:
\begin{lstlisting}
$ find ~/dir1 ~/dir2 ~/dir3  -name "*.fasta" |\
  parallel -a - echo {/}
\end{lstlisting}
output:
\begin{lstlisting}
seq1.fasta
seq2.fasta
alnfile.fasta
test_project.fasta
hs_owlmonkey.fasta
genomic-seq.fasta
testaln.fasta
test.fasta
\end{lstlisting}

copy all fasta to the current working directory:

\begin{lstlisting}
$ ls -l *.fasta

ls: cannot access *.fasta: No such file or directory

$ find ~/tmp/ /home/lindenb/daily/  -name "*.fasta" |\
   parallel -a - cp  {} {/}  && \
   ls -l *.fasta
\end{lstlisting}
output:
\begin{lstlisting}
-rw------- 1 lindenb lindenb    294 Oct  3 10:22 alnfile.fasta
-rw------- 1 lindenb lindenb 171524 Oct  3 10:22 genomic-seq.fasta
-rw------- 1 lindenb lindenb    416 Oct  3 10:22 hs_owlmonkey.fasta
-rw------- 1 lindenb lindenb   7194 Oct  3 10:22 seq1.fasta
-rw------- 1 lindenb lindenb  25756 Oct  3 10:22 seq2.fasta
-rw------- 1 lindenb lindenb   4620 Oct  3 10:22 testaln.fasta
-rw------- 1 lindenb lindenb    804 Oct  3 10:22 test.fasta
-rw------- 1 lindenb lindenb   3475 Oct  3 10:22 test_project.fasta
\end{lstlisting}

The replacement string \texttt{'\{/\}'} can be replaced with \texttt{'--basenamereplace'}:
\begin{lstlisting}
$ find ~/tmp/ /home/lindenb/daily/  -name "*.fasta" |\
   parallel --basenamereplace BASE_FILE_NAME -a - cp {}  BASE_FILE_NAME && \
   ls -l *.fasta
\end{lstlisting}
output:
\begin{lstlisting}
-rw------- 1 lindenb lindenb    294 Oct  3 11:34 alnfile.fasta
-rw------- 1 lindenb lindenb 171524 Oct  3 11:34 genomic-seq.fasta
-rw------- 1 lindenb lindenb    416 Oct  3 11:34 hs_owlmonkey.fasta
-rw------- 1 lindenb lindenb   7194 Oct  3 11:34 seq1.fasta
-rw------- 1 lindenb lindenb  25756 Oct  3 11:34 seq2.fasta
-rw------- 1 lindenb lindenb   4620 Oct  3 11:34 testaln.fasta
-rw------- 1 lindenb lindenb    804 Oct  3 11:34 test.fasta
-rw------- 1 lindenb lindenb   3475 Oct  3 11:34 test_project.fasta
\end{lstlisting}


The replacement string  \texttt{\{/.\}} removes the path and the extension.\\


Sorting the BAM in the current working directory:
\begin{lstlisting}
$ ls -l sorted_*.bam

ls: cannot access sorted_*.bam: No such file or directory

$ find ./ -name "*.bam" |\
  parallel -a - '../samtools sort' {} sorted_{/.} && \
  ls -l sorted_*.bam
\end{lstlisting}
output:
\begin{lstlisting}
-rw-rw-r-- 1 lindenb lindenb 126888 Oct  3 10:31 sorted_ex1a.bam
-rw-rw-r-- 1 lindenb lindenb 126583 Oct  3 10:31 sorted_ex1.bam
-rw-rw-r-- 1 lindenb lindenb 126878 Oct  3 10:31 sorted_ex1b.bam
-rw-rw-r-- 1 lindenb lindenb 208594 Oct  3 10:31 sorted_ex1f.bam
-rw-rw-r-- 1 lindenb lindenb 180639 Oct  3 10:31 sorted_ex1f-rmduppe.bam
-rw-rw-r-- 1 lindenb lindenb 132225 Oct  3 10:31 sorted_ex1f-rmdupse.bam
-rw-rw-r-- 1 lindenb lindenb    478 Oct  3 10:31 sorted_toy.bam
\end{lstlisting}


The replacement string \texttt{'\{//\}'} keeps only the path:
\begin{lstlisting}
find dir -name "*.fa" |\
 parallel echo {//} |\
 sort | uniq
\end{lstlisting}
output:
\begin{lstlisting}
dir/dir1
dir/dir2
dir/dir3
\end{lstlisting}
The replacement string \texttt{'\{//\}'} can be changed with --dirnamereplace.
\begin{lstlisting}
find dir -name "*.fa" |\
 parallel --dirnamereplace BARBALALA echo BARBALALA |\
 sort | uniq
\end{lstlisting}
output:
\begin{lstlisting}
dir/dir1
dir/dir2
dir/dir3
\end{lstlisting}


The replacement string \texttt{\{\#\}} gives the job number:
\begin{lstlisting}
$ find ~/tmp -name "*.fasta" |\
  parallel -a - echo {#} {/}
\end{lstlisting}
output:
\begin{lstlisting}
1 seq1.fasta
2 seq2.fasta
4 test_project.fasta
3 alnfile.fasta
5 hs_owlmonkey.fasta
6 genomic-seq.fasta
7 testaln.fasta
8 test.fasta
\end{lstlisting}

The replacement string \texttt{\{\#\}} can be changed with --seqreplace:
\begin{lstlisting}
$ find ~/tmp -name "*.fasta" | parallel -a - --seqreplace JOBNUM echo JOBNUM {/}
\end{lstlisting}
output:
\begin{lstlisting}
1 seq1.fasta
2 seq2.fasta
3 alnfile.fasta
4 test_project.fasta
5 hs_owlmonkey.fasta
6 genomic-seq.fasta
7 testaln.fasta
8 test.fasta
\end{lstlisting}


%% for I in 01 02 03 04 05; do ../misc/wgsim ex1.fa -N 1000 ${I}_F.fastq ${I}_R.fastq && gzip --best *.fastq ; done
\subsection{Positional replacement strings}
With multiple input sources the argument from the individual input sources can be access with \{number\}:


The positional replacement strings can also be modified using  \texttt{'/'}  or \texttt{'//'} or \texttt{'/.'} or  \texttt{'.'}:

aligning with 'BWA aln' two pairs of fastqs on two indexed references.
\begin{lstlisting}
$ parallel  bwa aln -f {1//}/{2/.}_{1/.}.sai {2} {1} \
 ::: examples/01_F.fastq.gz examples/01_R.fastq.gz examples/02_F.fastq.gz  examples/02_R.fastq.gz \
 ::: examples/toy.fa examples/ex1.fa
\end{lstlisting}
will generate
\begin{lstlisting}
bwa aln -f examples/toy_01_F.fastq.sai examples/toy.fa examples/01_F.fastq.gz
bwa aln -f examples/ex1_01_F.fastq.sai examples/ex1.fa examples/01_F.fastq.gz
bwa aln -f examples/toy_01_R.fastq.sai examples/toy.fa examples/01_R.fastq.gz
bwa aln -f examples/ex1_01_R.fastq.sai examples/ex1.fa examples/01_R.fastq.gz
bwa aln -f examples/toy_02_F.fastq.sai examples/toy.fa examples/02_F.fastq.gz
bwa aln -f examples/ex1_02_F.fastq.sai examples/ex1.fa examples/02_F.fastq.gz
bwa aln -f examples/toy_02_R.fastq.sai examples/toy.fa examples/02_R.fastq.gz
bwa aln -f examples/ex1_02_R.fastq.sai examples/ex1.fa examples/02_R.fastq.gz
\end{lstlisting}

output:
\begin{lstlisting}
$ ls -lah examples/*.sai

-rw-rw-r-- 1 lindenb lindenb  27K Oct  3 12:38 examples/ex1_01_F.fastq.sai
-rw-rw-r-- 1 lindenb lindenb  27K Oct  3 12:38 examples/ex1_01_R.fastq.sai
-rw-rw-r-- 1 lindenb lindenb  27K Oct  3 12:38 examples/ex1_02_F.fastq.sai
-rw-rw-r-- 1 lindenb lindenb  27K Oct  3 12:38 examples/ex1_02_R.fastq.sai
-rw-rw-r-- 1 lindenb lindenb 4.0K Oct  3 12:38 examples/toy_01_F.fastq.sai
-rw-rw-r-- 1 lindenb lindenb 4.0K Oct  3 12:38 examples/toy_01_R.fastq.sai
-rw-rw-r-- 1 lindenb lindenb 4.0K Oct  3 12:38 examples/toy_02_F.fastq.sai
-rw-rw-r-- 1 lindenb lindenb 4.0K Oct  3 12:38 examples/toy_02_R.fastq.sai
\end{lstlisting}

\subsection{Input from columns}

The columns in a file can be bound to positional replacement strings using --colsep. Here the columns are separated with TAB:

\begin{lstlisting}
$ find examples/ -name "*.fastq.gz" |\
  sort |\
  paste -- - - |\
  parallel  --colsep '\t' bwa mem examples/ex1.fa {1} {2} ">" {1//}/{1/.}_{2/.}.sam
\end{lstlisting}
will generate
\begin{lstlisting}
bwa mem examples/ex1.fa examples/01_F.fastq.gz examples/01_R.fastq.gz > examples/01_F.fastq_01_R.fastq.sam
bwa mem examples/ex1.fa examples/02_F.fastq.gz examples/02_R.fastq.gz > examples/02_F.fastq_02_R.fastq.sam
bwa mem examples/ex1.fa examples/03_F.fastq.gz examples/03_R.fastq.gz > examples/03_F.fastq_03_R.fastq.sam
bwa mem examples/ex1.fa examples/04_F.fastq.gz examples/04_R.fastq.gz > examples/04_F.fastq_04_R.fastq.sam
bwa mem examples/ex1.fa examples/05_F.fastq.gz examples/05_R.fastq.gz > examples/05_F.fastq_05_R.fastq.sam
\end{lstlisting}
output:
\begin{lstlisting}
$ ls -la examples/*.sam

-rw-rw-r-- 1 lindenb lindenb 447025 Oct  3 13:03 examples/01_F.fastq_01_R.fastq.sam
-rw-rw-r-- 1 lindenb lindenb 447025 Oct  3 13:03 examples/02_F.fastq_02_R.fastq.sam
-rw-rw-r-- 1 lindenb lindenb 447025 Oct  3 13:03 examples/03_F.fastq_03_R.fastq.sam
-rw-rw-r-- 1 lindenb lindenb 447025 Oct  3 13:03 examples/04_F.fastq_04_R.fastq.sam
-rw-rw-r-- 1 lindenb lindenb 447025 Oct  3 13:03 examples/05_F.fastq_05_R.fastq.sam
\end{lstlisting}


\subsection{Header defined replacement strings}
With --header GNU Parallel will use the first value of the input source as the name of the replacement string. Only the non-modified version {} is supported
\begin{lstlisting}
$ parallel --header : primer3-2.3.5/src/ntdpal {FORWARD} {REVERSE} g \
   ::: REVERSE ATCTGACTCGTGC ACTGATCGATCGATCG \
   ::: FORWARD ATAGTAATAT ACTATA GAAATTC
\end{lstlisting}
output:
\begin{lstlisting}
|ATAGTAATAT|  |ATCTGACTCGTGC| g score=2.00 len=4 |6,0|7,1|8,2|9,3|
|ACTATA|  |ATCTGACTCGTGC| g score=1.00 len=6 |0,5|1,6|2,7|3,9|4,10|5,11|
|GAAATTC|  |ATCTGACTCGTGC| g score=2.00 len=3 |3,0|5,1|6,2|
|ATAGTAATAT|  |ACTGATCGATCGATCG| g score=2.00 len=8 |0,4|1,5|2,6|3,7|6,8|7,9|8,12|9,13|
|ACTATA|  |ACTGATCGATCGATCG| g score=3.00 len=6 |0,0|1,1|2,2|3,4|4,5|5,8|
|GAAATTC|  |ACTGATCGATCGATCG| g score=1.00 len=6 |0,7|1,8|2,11|3,12|5,13|6,14|
\end{lstlisting}

\subsection{More than one argument}
With --xargs will GNU Parallel fit as many arguments as possible on a single line:
\begin{lstlisting}
$ seq 1 100000 | parallel --xargs echo  | wc -l
5
\end{lstlisting}
The 100000 arguments fitted on 5 lines.

The maximal length of a single line can be set with -s. With a maximal line length of 10000 chars 595 commands will be run:
\begin{lstlisting}
$ seq 1 100000 | parallel --xargs -s  1000 echo  | wc -l
595
\end{lstlisting}

\subsection{Quoting}
Command lines that contain special characters may need to be protected from the shell.
\begin{lstlisting}
 find dir1 -name "*.fa" |\
 xargs awk '/^>/ {printf("\n%s\t",$0);next;} { printf("%s",$0);} END { printf("\n");} ' |\
 head | cut -c 1-100

>seq1	CACTAGTGGCTCATTGTAAATGTGTGGTTTAACTCGTCCATGGCCCAGCATTAGGGAGCTGTGGACCCTGCAGCCTGGCTGTGGGGGCCGCAGT
>seq2	TTCAAATGAACTTCTGTAATTGAAAAATTCATTTAAGAAATTACAAAATATAGTTGAAAGCTCTAACAATAGACTAAACCAAGCAGAAGAAAGA
>ref	AGCATGTTAGATAAGATAGCTGTGCTAGTAGGCAGTCAGCGCCAT
>ref2	aggttttataaaacaattaagtctacagagcaactacgcg
\end{lstlisting}

this won't work:
\begin{lstlisting}
Command lines that contain special characters may need to be protected from the shell.
\begin{lstlisting}
 find dir1 -name "*.fa" |\
 parallel awk '/^>/ {printf("\n%s\t",$0);next;} { printf("%s",$0);} END { printf("\n");} ' 
\end{lstlisting}
To quote the command use \texttt{'-q'}:
\begin{lstlisting}
 find dir1 -name "*.fa" |\
 parallel -q awk '/^>/ {printf("\n%s\t",$0);next;} { printf("%s",$0);} END { printf("\n");} ' |\
 head | cut -c 1-100

>seq1	CACTAGTGGCTCATTGTAAATGTGTGGTTTAACTCGTCCATGGCCCAGCATTAGGGAGCTGTGGACCCTGCAGCCTGGCTGTGGGGGCCGCAGT
>seq2	TTCAAATGAACTTCTGTAATTGAAAAATTCATTTAAGAAATTACAAAATATAGTTGAAAGCTCTAACAATAGACTAAACCAAGCAGAAGAAAGA

>ref	AGCATGTTAGATAAGATAGCTGTGCTAGTAGGCAGTCAGCGCCAT
>ref2	aggttttataaaacaattaagtctacagagcaactacgcg
\end{lstlisting}

Or you can quote the critical part using \texttt{\textbackslash{}'}:
\begin{lstlisting}
$ find dir -name "*.fa" |\
 parallel awk \' '/^>/ {printf("\n%s\t",$0);next;} { printf("%s",$0);} END { printf("\n");} ' \' |\
 head | cut -c 1-100

>seq1	CACTAGTGGCTCATTGTAAATGTGTGGTTTAACTCGTCCATGGCCCAGCATTAGGGAGCTGTGGACCCTGCAGCCTGGCTGTGGGGGCCGCAGT
>seq2	TTCAAATGAACTTCTGTAATTGAAAAATTCATTTAAGAAATTACAAAATATAGTTGAAAGCTCTAACAATAGACTAAACCAAGCAGAAGAAAGA

>ref	AGCATGTTAGATAAGATAGCTGTGCTAGTAGGCAGTCAGCGCCAT
>ref2	aggttttataaaacaattaagtctacagagcaactacgcg
\end{lstlisting}

%% missing example for   parallel --shellquote

\subsection{Trimming space}
Space can be trimmed on the arguments using  \texttt{--trim}
\begin{lstlisting}
$ parallel --trim l echo [{}] :::  " A " " T " " C " " G "
\end{lstlisting}
output:
\begin{lstlisting}
[A ]
[T ]
[C ]
[G ]
\end{lstlisting}

\begin{lstlisting}
$ parallel --trim r echo [{}] :::  " A " " T " " C " " G "
\end{lstlisting}
output:
\begin{lstlisting}
[ A]
[ T]
[ C]
[ G]
\end{lstlisting}

\begin{lstlisting}
$ parallel --trim lr echo [{}] :::  " A " " T " " C " " G "
\end{lstlisting}
output:
\begin{lstlisting}
[A]
[T]
[C]
[G]
\end{lstlisting}

\section{Controling the output}
The output can prefixed with the argument:
\begin{lstlisting}
$ parallel --tag echo prefix-{} ::: A T G C
\end{lstlisting}
output:
\begin{lstlisting}Output (the order may be different):
A	prefix-A
T	prefix-T
G	prefix-G
C	prefix-C
\end{lstlisting}

To prefix it with another string use --tagstring:
\begin{lstlisting}
parallel --tagstring suffix-{} echo  ::: A T G C
\end{lstlisting}
output:
\begin{lstlisting}
suffix-A	A
suffix-T	T
suffix-G	G
suffix-C	C
\end{lstlisting}

\subsection{To see what commands will be run without running them}
\begin{lstlisting}
$ find ./ -name "*.fastq.gz" |\
	parallel --dryrun  bwa aln -f {/.}.sai toy.fa {}
\end{lstlisting}
output:
\begin{lstlisting}
bwa aln -f 02_R_.fastq.sai toy.fa ./02_R_.fastq.gz
bwa aln -f 03_R_.fastq.sai toy.fa ./03_R_.fastq.gz
bwa aln -f 01_R_.fastq.sai toy.fa ./01_R_.fastq.gz
bwa aln -f 05_F_.fastq.sai toy.fa ./05_F_.fastq.gz
bwa aln -f 05_R_.fastq.sai toy.fa ./05_R_.fastq.gz
bwa aln -f 04_R_.fastq.sai toy.fa ./04_R_.fastq.gz
bwa aln -f 03_F_.fastq.sai toy.fa ./03_F_.fastq.gz
bwa aln -f 04_F_.fastq.sai toy.fa ./04_F_.fastq.gz
bwa aln -f 01_F_.fastq.sai toy.fa ./01_F_.fastq.gz
bwa aln -f 02_F_.fastq.sai toy.fa ./02_F_.fastq.gz
\end{lstlisting}

\subsection{To print the command before running them use --verbose}
\begin{lstlisting}
$ find ./ -name "*.fastq.gz" |\
	parallel --verbose  bwa aln -f {/.}.sai toy.fa {}
\end{lstlisting}
output:
\begin{lstlisting}
bwa aln -f 02_R_.fastq.sai toy.fa ./02_R_.fastq.gz
bwa aln -f 03_R_.fastq.sai toy.fa ./03_R_.fastq.gz
[bwa_aln] 17bp reads: max_diff = 2
[bwa_aln] 38bp reads: max_diff = 3
[bwa_aln] 64bp reads: max_diff = 4
[bwa_aln] 93bp reads: max_diff = 5
[bwa_aln] 124bp reads: max_diff = 6
[bwa_aln] 157bp reads: max_diff = 7
[bwa_aln] 190bp reads: max_diff = 8
[bwa_aln] 225bp reads: max_diff = 9
[bwa_aln_core] calculate SA coordinate... 0.01 sec
[bwa_aln_core] write to the disk... 0.00 sec
[bwa_aln_core] 1000 sequences have been processed.
[main] Version: 0.7.4-r385
[main] CMD: bwa aln -f 02_R_.fastq.sai toy.fa ./02_R_.fastq.gz
[main] Real time: 0.017 sec; CPU: 0.020 sec
bwa aln -f 01_R_.fastq.sai toy.fa ./01_R_.fastq.gz
[bwa_aln] 17bp reads: max_diff = 2
[bwa_aln] 38bp reads: max_diff = 3
[bwa_aln] 64bp reads: max_diff = 4
[bwa_aln] 93bp reads: max_diff = 5
[bwa_aln] 124bp reads: max_diff = 6
[bwa_aln] 157bp reads: max_diff = 7
[bwa_aln] 190bp reads: max_diff = 8
[bwa_aln] 225bp reads: max_diff = 9
[bwa_aln_core] calculate SA coordinate... 0.00 sec
[bwa_aln_core] write to the disk... 0.00 sec
[bwa_aln_core] 1000 sequences have been processed.
[main] Version: 0.7.4-r385
[main] CMD: bwa aln -f 03_R_.fastq.sai toy.fa ./03_R_.fastq.gz
[main] Real time: 0.016 sec; CPU: 0.012 sec
bwa aln -f 05_F_.fastq.sai toy.fa ./05_F_.fastq.gz
[bwa_aln] 17bp reads: max_diff = 2
[bwa_aln] 38bp reads: max_diff = 3
[bwa_aln] 64bp reads: max_diff = 4
[bwa_aln] 93bp reads: max_diff = 5
[bwa_aln] 124bp reads: max_diff = 6
[bwa_aln] 157bp reads: max_diff = 7
[bwa_aln] 190bp reads: max_diff = 8
[bwa_aln] 225bp reads: max_diff = 9
[bwa_aln_core] calculate SA coordinate... 0.01 sec
[bwa_aln_core] write to the disk... 0.00 sec
[bwa_aln_core] 1000 sequences have been processed.
[main] Version: 0.7.4-r385
[main] CMD: bwa aln -f 01_R_.fastq.sai toy.fa ./01_R_.fastq.gz
[main] Real time: 0.012 sec; CPU: 0.012 sec
bwa aln -f 05_R_.fastq.sai toy.fa ./05_R_.fastq.gz
[bwa_aln] 17bp reads: max_diff = 2
[bwa_aln] 38bp reads: max_diff = 3
[bwa_aln] 64bp reads: max_diff = 4
[bwa_aln] 93bp reads: max_diff = 5
[bwa_aln] 124bp reads: max_diff = 6
[bwa_aln] 157bp reads: max_diff = 7
[bwa_aln] 190bp reads: max_diff = 8
[bwa_aln] 225bp reads: max_diff = 9
[bwa_aln_core] calculate SA coordinate... 0.00 sec
[bwa_aln_core] write to the disk... 0.00 sec
[bwa_aln_core] 1000 sequences have been processed.
[main] Version: 0.7.4-r385
[main] CMD: bwa aln -f 05_F_.fastq.sai toy.fa ./05_F_.fastq.gz
[main] Real time: 0.013 sec; CPU: 0.012 sec
bwa aln -f 04_R_.fastq.sai toy.fa ./04_R_.fastq.gz
[bwa_aln] 17bp reads: max_diff = 2
[bwa_aln] 38bp reads: max_diff = 3
[bwa_aln] 64bp reads: max_diff = 4
[bwa_aln] 93bp reads: max_diff = 5
[bwa_aln] 124bp reads: max_diff = 6
[bwa_aln] 157bp reads: max_diff = 7
[bwa_aln] 190bp reads: max_diff = 8
[bwa_aln] 225bp reads: max_diff = 9
[bwa_aln_core] calculate SA coordinate... 0.01 sec
[bwa_aln_core] write to the disk... 0.00 sec
[bwa_aln_core] 1000 sequences have been processed.
[main] Version: 0.7.4-r385
[main] CMD: bwa aln -f 05_R_.fastq.sai toy.fa ./05_R_.fastq.gz
[main] Real time: 0.012 sec; CPU: 0.012 sec
bwa aln -f 03_F_.fastq.sai toy.fa ./03_F_.fastq.gz
[bwa_aln] 17bp reads: max_diff = 2
[bwa_aln] 38bp reads: max_diff = 3
[bwa_aln] 64bp reads: max_diff = 4
[bwa_aln] 93bp reads: max_diff = 5
[bwa_aln] 124bp reads: max_diff = 6
[bwa_aln] 157bp reads: max_diff = 7
[bwa_aln] 190bp reads: max_diff = 8
[bwa_aln] 225bp reads: max_diff = 9
[bwa_aln_core] calculate SA coordinate... 0.01 sec
[bwa_aln_core] write to the disk... 0.00 sec
[bwa_aln_core] 1000 sequences have been processed.
[main] Version: 0.7.4-r385
[main] CMD: bwa aln -f 04_R_.fastq.sai toy.fa ./04_R_.fastq.gz
[main] Real time: 0.012 sec; CPU: 0.012 sec
bwa aln -f 04_F_.fastq.sai toy.fa ./04_F_.fastq.gz
[bwa_aln] 17bp reads: max_diff = 2
[bwa_aln] 38bp reads: max_diff = 3
[bwa_aln] 64bp reads: max_diff = 4
[bwa_aln] 93bp reads: max_diff = 5
[bwa_aln] 124bp reads: max_diff = 6
[bwa_aln] 157bp reads: max_diff = 7
[bwa_aln] 190bp reads: max_diff = 8
[bwa_aln] 225bp reads: max_diff = 9
[bwa_aln_core] calculate SA coordinate... 0.00 sec
[bwa_aln_core] write to the disk... 0.00 sec
[bwa_aln_core] 1000 sequences have been processed.
[main] Version: 0.7.4-r385
[main] CMD: bwa aln -f 03_F_.fastq.sai toy.fa ./03_F_.fastq.gz
[main] Real time: 0.013 sec; CPU: 0.012 sec
bwa aln -f 02_F_.fastq.sai toy.fa ./02_F_.fastq.gz
[bwa_aln] 17bp reads: max_diff = 2
[bwa_aln] 38bp reads: max_diff = 3
[bwa_aln] 64bp reads: max_diff = 4
[bwa_aln] 93bp reads: max_diff = 5
[bwa_aln] 124bp reads: max_diff = 6
[bwa_aln] 157bp reads: max_diff = 7
[bwa_aln] 190bp reads: max_diff = 8
[bwa_aln] 225bp reads: max_diff = 9
[bwa_aln_core] calculate SA coordinate... 0.00 sec
[bwa_aln_core] write to the disk... 0.00 sec
[bwa_aln_core] 1000 sequences have been processed.
[main] Version: 0.7.4-r385
[main] CMD: bwa aln -f 04_F_.fastq.sai toy.fa ./04_F_.fastq.gz
[main] Real time: 0.012 sec; CPU: 0.012 sec
bwa aln -f 01_F_.fastq.sai toy.fa ./01_F_.fastq.gz
[bwa_aln] 17bp reads: max_diff = 2
[bwa_aln] 38bp reads: max_diff = 3
[bwa_aln] 64bp reads: max_diff = 4
[bwa_aln] 93bp reads: max_diff = 5
[bwa_aln] 124bp reads: max_diff = 6
[bwa_aln] 157bp reads: max_diff = 7
[bwa_aln] 190bp reads: max_diff = 8
[bwa_aln] 225bp reads: max_diff = 9
[bwa_aln_core] calculate SA coordinate... 0.00 sec
[bwa_aln_core] write to the disk... 0.00 sec
[bwa_aln_core] 1000 sequences have been processed.
[main] Version: 0.7.4-r385
[main] CMD: bwa aln -f 02_F_.fastq.sai toy.fa ./02_F_.fastq.gz
[main] Real time: 0.016 sec; CPU: 0.012 sec
[bwa_aln] 17bp reads: max_diff = 2
[bwa_aln] 38bp reads: max_diff = 3
[bwa_aln] 64bp reads: max_diff = 4
[bwa_aln] 93bp reads: max_diff = 5
[bwa_aln] 124bp reads: max_diff = 6
[bwa_aln] 157bp reads: max_diff = 7
[bwa_aln] 190bp reads: max_diff = 8
[bwa_aln] 225bp reads: max_diff = 9
[bwa_aln_core] calculate SA coordinate... 0.01 sec
[bwa_aln_core] write to the disk... 0.00 sec
[bwa_aln_core] 1000 sequences have been processed.
[main] Version: 0.7.4-r385
[main] CMD: bwa aln -f 01_F_.fastq.sai toy.fa ./01_F_.fastq.gz
[main] Real time: 0.012 sec; CPU: 0.012 sec
\end{lstlisting}

\subsection{GNU Parallel will postpone the output until the command completes}
\begin{lstlisting}
$ parallel -j 3 --verbose 'bwa aln -f 0{}_F.fastq.sai toy.fa 0{}_F_.fastq.gz ; sleep {} ; bwa aln -f 0{}_R.fastq.sai toy.fa 0{}_R_.fastq.gz ' ::: 1 2 3 4 5
\end{lstlisting}
output:
\begin{lstlisting}
$ parallel -j 3 --verbose  'bwa aln -f 0{}_F.fastq.sai toy.fa 0{}_F_.fastq.gz ; sleep {} ; bwa aln -f 0{}_R.fastq.sai toy.fa 0{}_R_.fastq.gz ' ::: 1 2 3 4 5

bwa aln -f 01_F.fastq.sai toy.fa 01_F_.fastq.gz ; sleep 1 ; bwa aln -f 01_R.fastq.sai toy.fa 01_R_.fastq.gz 
bwa aln -f 02_F.fastq.sai toy.fa 02_F_.fastq.gz ; sleep 2 ; bwa aln -f 02_R.fastq.sai toy.fa 02_R_.fastq.gz 
bwa aln -f 03_F.fastq.sai toy.fa 03_F_.fastq.gz ; sleep 3 ; bwa aln -f 03_R.fastq.sai toy.fa 03_R_.fastq.gz 
(...)
[main] CMD: bwa aln -f 01_F.fastq.sai toy.fa 01_F_.fastq.gz
(...)
[main] CMD: bwa aln -f 01_R.fastq.sai toy.fa 01_R_.fastq.gz
(...)
bwa aln -f 04_F.fastq.sai toy.fa 04_F_.fastq.gz ; sleep 4 ; bwa aln -f 04_R.fastq.sai toy.fa 04_R_.fastq.gz 
(...)
[main] CMD: bwa aln -f 02_F.fastq.sai toy.fa 02_F_.fastq.gz
(...)
[main] CMD: bwa aln -f 02_R.fastq.sai toy.fa 02_R_.fastq.gz
(...)
bwa aln -f 05_F.fastq.sai toy.fa 05_F_.fastq.gz ; sleep 5 ; bwa aln -f 05_R.fastq.sai toy.fa 05_R_.fastq.gz 
(...)
[main] CMD: bwa aln -f 03_F.fastq.sai toy.fa 03_F_.fastq.gz
(...)
[main] CMD: bwa aln -f 03_R.fastq.sai toy.fa 03_R_.fastq.gz
(...)
[main] CMD: bwa aln -f 04_F.fastq.sai toy.fa 04_F_.fastq.gz
(...)
[main] CMD: bwa aln -f 04_R.fastq.sai toy.fa 04_R_.fastq.gz
(...)
[main] CMD: bwa aln -f 05_F.fastq.sai toy.fa 05_F_.fastq.gz
(...)
[main] CMD: bwa aln -f 05_R.fastq.sai toy.fa 05_R_.fastq.gz
\end{lstlisting}



\subsection{To get the output immediately use --ungroup}
\begin{lstlisting}
$ parallel -j 3 --verbose --ungroup 'bwa aln -f 0{}_F.fastq.sai toy.fa 0{}_F_.fastq.gz ; sleep {} ; bwa aln -f 0{}_R.fastq.sai toy.fa 0{}_R_.fastq.gz ' ::: 1 2 3 4 5
\end{lstlisting}
output:
\begin{lstlisting}
bwa aln -f 01_F.fastq.sai toy.fa 01_F_.fastq.gz ; sleep 1 ; bwa aln -f 01_R.fastq.sai toy.fa 01_R_.fastq.gz 
bwa aln -f 02_F.fastq.sai toy.fa 02_F_.fastq.gz ; sleep 2 ; bwa aln -f 02_R.fastq.sai toy.fa 02_R_.fastq.gz 
(...)
bwa aln -f 03_F.fastq.sai toy.fa 03_F_.fastq.gz ; sleep 3 ; bwa aln -f 03_R.fastq.sai toy.fa 03_R_.fastq.gz 
(...)
[main] CMD: bwa aln -f 01_F.fastq.sai toy.fa 01_F_.fastq.gz
(...)
[main] CMD: bwa aln -f 02_F.fastq.sai toy.fa 02_F_.fastq.gz
(...)
[main] CMD: bwa aln -f 03_F.fastq.sai toy.fa 03_F_.fastq.gz
(...)
[main] CMD: bwa aln -f 01_R.fastq.sai toy.fa 01_R_.fastq.gz
(...)
bwa aln -f 04_F.fastq.sai toy.fa 04_F_.fastq.gz ; sleep 4 ; bwa aln -f 04_R.fastq.sai toy.fa 04_R_.fastq.gz 
(...)
[main] CMD: bwa aln -f 04_F.fastq.sai toy.fa 04_F_.fastq.gz
(...)
[main] CMD: bwa aln -f 02_R.fastq.sai toy.fa 02_R_.fastq.gz
(...)
bwa aln -f 05_F.fastq.sai toy.fa 05_F_.fastq.gz ; sleep 5 ; bwa aln -f 05_R.fastq.sai toy.fa 05_R_.fastq.gz 
(...)
[main] CMD: bwa aln -f 05_F.fastq.sai toy.fa 05_F_.fastq.gz
(...)
[main] CMD: bwa aln -f 03_R.fastq.sai toy.fa 03_R_.fastq.gz
(...)
[main] CMD: bwa aln -f 04_R.fastq.sai toy.fa 04_R_.fastq.gz
(...)
[main] CMD: bwa aln -f 05_R.fastq.sai toy.fa 05_R_.fastq.gz
(...)
\end{lstlisting}

--ungroup is fast, but can cause half a line from one job to be mixed with half a line of another job. That has happend in the second line, where the line '4-middle' is mixed with '2-start'.

To avoid this use --linebuffer (which, however, is much slower):
\begin{lstlisting}
$ parallel -j 3 --verbose --linebuffer 'bwa aln -f 0{}_F.fastq.sai toy.fa 0{}_F_.fastq.gz ; sleep {} ; bwa aln -f 0{}_R.fastq.sai toy.fa 0{}_R_.fastq.gz ' ::: 1 2 3 4 5
\end{lstlisting}
output:
\begin{lstlisting}
bwa aln -f 01_F.fastq.sai toy.fa 01_F_.fastq.gz ; sleep 1 ; bwa aln -f 01_R.fastq.sai toy.fa 01_R_.fastq.gz 
bwa aln -f 02_F.fastq.sai toy.fa 02_F_.fastq.gz ; sleep 2 ; bwa aln -f 02_R.fastq.sai toy.fa 02_R_.fastq.gz 
bwa aln -f 03_F.fastq.sai toy.fa 03_F_.fastq.gz ; sleep 3 ; bwa aln -f 03_R.fastq.sai toy.fa 03_R_.fastq.gz 
(...)
[main] CMD: bwa aln -f 01_F.fastq.sai toy.fa 01_F_.fastq.gz
(...)
[main] CMD: bwa aln -f 03_F.fastq.sai toy.fa 03_F_.fastq.gz
(...)
[main] CMD: bwa aln -f 02_F.fastq.sai toy.fa 02_F_.fastq.gz
(...)
[main] CMD: bwa aln -f 01_R.fastq.sai toy.fa 01_R_.fastq.gz
(...)
bwa aln -f 04_F.fastq.sai toy.fa 04_F_.fastq.gz ; sleep 4 ; bwa aln -f 04_R.fastq.sai toy.fa 04_R_.fastq.gz 
(...)
[main] CMD: bwa aln -f 04_F.fastq.sai toy.fa 04_F_.fastq.gz
(...)
[main] CMD: bwa aln -f 02_R.fastq.sai toy.fa 02_R_.fastq.gz
(...)
bwa aln -f 05_F.fastq.sai toy.fa 05_F_.fastq.gz ; sleep 5 ; bwa aln -f 05_R.fastq.sai toy.fa 05_R_.fastq.gz 
(...)
[main] CMD: bwa aln -f 05_F.fastq.sai toy.fa 05_F_.fastq.gz
(...)
[main] CMD: bwa aln -f 03_R.fastq.sai toy.fa 03_R_.fastq.gz
(...)
[main] CMD: bwa aln -f 04_R.fastq.sai toy.fa 04_R_.fastq.gz
(...)
[main] CMD: bwa aln -f 05_R.fastq.sai toy.fa 05_R_.fastq.gz
(...)
\end{lstlisting}

To force the output in the same order as the arguments use --keep-order/-k:

\begin{lstlisting}
$ parallel -j 2 --verbose --keep-order 'bwa aln -f 0{}_F.fastq.sai toy.fa 0{}_F_.fastq.gz ; sleep {} ; bwa aln -f 0{}_R.fastq.sai toy.fa 0{}_R_.fastq.gz ' ::: 1 2 3 4 5
\end{lstlisting}
output:
\begin{lstlisting}
bwa aln -f 01_F.fastq.sai toy.fa 01_F_.fastq.gz ; sleep 1 ; bwa aln -f 01_R.fastq.sai toy.fa 01_R_.fastq.gz 
bwa aln -f 02_F.fastq.sai toy.fa 02_F_.fastq.gz ; sleep 2 ; bwa aln -f 02_R.fastq.sai toy.fa 02_R_.fastq.gz 
(...)
[main] CMD: bwa aln -f 01_F.fastq.sai toy.fa 01_F_.fastq.gz
(...)
[main] CMD: bwa aln -f 01_R.fastq.sai toy.fa 01_R_.fastq.gz
(...)
bwa aln -f 03_F.fastq.sai toy.fa 03_F_.fastq.gz ; sleep 3 ; bwa aln -f 03_R.fastq.sai toy.fa 03_R_.fastq.gz 
(...)
[main] CMD: bwa aln -f 02_F.fastq.sai toy.fa 02_F_.fastq.gz
(...)
[main] CMD: bwa aln -f 02_R.fastq.sai toy.fa 02_R_.fastq.gz
(...)
bwa aln -f 04_F.fastq.sai toy.fa 04_F_.fastq.gz ; sleep 4 ; bwa aln -f 04_R.fastq.sai toy.fa 04_R_.fastq.gz 
(...)
[main] CMD: bwa aln -f 03_F.fastq.sai toy.fa 03_F_.fastq.gz
(...)
[main] CMD: bwa aln -f 03_R.fastq.sai toy.fa 03_R_.fastq.gz
(...)
bwa aln -f 05_F.fastq.sai toy.fa 05_F_.fastq.gz ; sleep 5 ; bwa aln -f 05_R.fastq.sai toy.fa 05_R_.fastq.gz 
(...)
[main] CMD: bwa aln -f 04_F.fastq.sai toy.fa 04_F_.fastq.gz
(...)
[main] CMD: bwa aln -f 04_R.fastq.sai toy.fa 04_R_.fastq.gz
(...)
[main] CMD: bwa aln -f 05_F.fastq.sai toy.fa 05_F_.fastq.gz
(...)
[main] CMD: bwa aln -f 05_R.fastq.sai toy.fa 05_R_.fastq.gz
\end{lstlisting}

\subsubsection{Saving output into files}
GNU Parallel can save the output of each job into files:
\begin{lstlisting}
$ parallel --files --verbose primer3-2.3.5/src/ntdpal {1} {2} g \
	::: AATCGTACGTACG ATAGCATCGA \
	::: AATCGTACGTCG  ATAGCATCGAG
\end{lstlisting}
output:
\begin{lstlisting}
/home/lindenb/package/primer3-2.3.5/src/ntdpal AATCGTACGTACG AATCGTACGTCG g
/home/lindenb/package/primer3-2.3.5/src/ntdpal AATCGTACGTACG ATAGCATCGAG g
/tmp/eLAmSSq0DY.par
/home/lindenb/package/primer3-2.3.5/src/ntdpal ATAGCATCGA AATCGTACGTCG g
/tmp/Yv7YvDPzzw.par
/home/lindenb/package/primer3-2.3.5/src/ntdpal ATAGCATCGA ATAGCATCGAG g
/tmp/mea6UUq1Li.par
/tmp/VLbG2gkq2c.par

$ head /tmp/eLAmSSq0DY.par /tmp/Yv7YvDPzzw.par /tmp/mea6UUq1Li.par /tmp/VLbG2gkq2c.par
==> /tmp/eLAmSSq0DY.par <==
|AATCGTACGTACG|  |AATCGTACGTCG| g score=11.00 len=12 |0,0|1,1|2,2|3,3|4,4|5,5|6,6|7,7|8,8|9,9|11,10|12,11|

==> /tmp/Yv7YvDPzzw.par <==
|AATCGTACGTACG|  |ATAGCATCGAG| g score=3.00 len=10 |1,0|2,1|3,2|4,3|5,4|6,5|7,7|8,8|10,9|12,10|

==> /tmp/mea6UUq1Li.par <==
|ATAGCATCGA|  |AATCGTACGTCG| g score=3.00 len=6 |4,0|5,1|6,2|7,3|8,4|9,6|

==> /tmp/VLbG2gkq2c.par <==
|ATAGCATCGA|  |ATAGCATCGAG| g score=10.00 len=10 |0,0|1,1|2,2|3,3|4,4|5,5|6,6|7,7|8,8|9,9|
\end{lstlisting}

By default GNU Parallel will cache the output in files in /tmp. This can be changed by setting \$TMPDIR or --tmpdir:
\begin{lstlisting}
$ parallel --files --verbose --tmpdir . primer3-2.3.5/src/ntdpal {1} {2} g \
	::: AATCGTACGTACG ATAGCATCGA \
	::: AATCGTACGTCG  ATAGCATCGAG
\end{lstlisting}
output:
\begin{lstlisting}
/home/lindenb/package/primer3-2.3.5/src/ntdpal AATCGTACGTACG AATCGTACGTCG g
/home/lindenb/package/primer3-2.3.5/src/ntdpal AATCGTACGTACG ATAGCATCGAG g
./0gCXxOB_fm.par
/home/lindenb/package/primer3-2.3.5/src/ntdpal ATAGCATCGA AATCGTACGTCG g
./SD67tLyZTs.par
/home/lindenb/package/primer3-2.3.5/src/ntdpal ATAGCATCGA ATAGCATCGAG g
./RR54nlC0yu.par
./K1c6EBsxZq.par
\end{lstlisting}

The output files can be saved in a structured way using --results:
\begin{lstlisting}
parallel --result ALN  bwa aln -f 0{}_R.fastq.sai toy.fa 0{}_R_.fastq.gz  ::: 1 2 3 4 5
\end{lstlisting}
output:
\begin{lstlisting}
$ find ALN
ALN
ALN/1
ALN/1/2
ALN/1/2/stdout
ALN/1/2/stderr
ALN/1/1
ALN/1/1/stdout
ALN/1/1/stderr
ALN/1/5
ALN/1/5/stdout
ALN/1/5/stderr
ALN/1/4
ALN/1/4/stdout
ALN/1/4/stderr
ALN/1/3
ALN/1/3/stdout
ALN/1/3/stderr
\end{lstlisting}

\begin{lstlisting}
 parallel --result NTDPAL  primer3-2.3.5/src/ntdpal {1} {2} g ::: AATCGTACGTACG ATAGCATCGA ::: AATCGTACGTCG  ATAGCATCGAG
\end{lstlisting}
output:
\begin{lstlisting}
$ find NTDPAL/
NTDPAL/
NTDPAL/1
NTDPAL/1/AATCGTACGTACG
NTDPAL/1/AATCGTACGTACG/2
NTDPAL/1/AATCGTACGTACG/2/ATAGCATCGAG
NTDPAL/1/AATCGTACGTACG/2/ATAGCATCGAG/stdout
NTDPAL/1/AATCGTACGTACG/2/ATAGCATCGAG/stderr
NTDPAL/1/AATCGTACGTACG/2/AATCGTACGTCG
NTDPAL/1/AATCGTACGTACG/2/AATCGTACGTCG/stdout
NTDPAL/1/AATCGTACGTACG/2/AATCGTACGTCG/stderr
NTDPAL/1/ATAGCATCGA
NTDPAL/1/ATAGCATCGA/2
NTDPAL/1/ATAGCATCGA/2/ATAGCATCGAG
NTDPAL/1/ATAGCATCGA/2/ATAGCATCGAG/stdout
NTDPAL/1/ATAGCATCGA/2/ATAGCATCGAG/stderr
NTDPAL/1/ATAGCATCGA/2/AATCGTACGTCG
NTDPAL/1/ATAGCATCGA/2/AATCGTACGTCG/stdout
NTDPAL/1/ATAGCATCGA/2/AATCGTACGTCG/stderr

$ cat NTDPAL/1/AATCGTACGTACG/2/ATAGCATCGAG/stdout
|AATCGTACGTACG|  |ATAGCATCGAG| g score=3.00 len=10 |1,0|2,1|3,2|4,3|5,4|6,5|7,7|8,8|10,9|12,10|
\end{lstlisting}


This is useful if you are running multiple variables:
\begin{lstlisting}
$ parallel --header : --result NTDPAL  primer3-2.3.5/src/ntdpal {1} {2} g \
	::: primer-foward AATCGTACGTACG ATAGCATCGA \
	::: primer-reverse AATCGTACGTCG  ATAGCATCGAG
\end{lstlisting}
Generated files:
\begin{lstlisting}
$ find NTDPAL/

NTDPAL/
NTDPAL/primer-foward
NTDPAL/primer-foward/AATCGTACGTACG
NTDPAL/primer-foward/AATCGTACGTACG/primer-reverse
NTDPAL/primer-foward/AATCGTACGTACG/primer-reverse/ATAGCATCGAG
NTDPAL/primer-foward/AATCGTACGTACG/primer-reverse/ATAGCATCGAG/stdout
NTDPAL/primer-foward/AATCGTACGTACG/primer-reverse/ATAGCATCGAG/stderr
NTDPAL/primer-foward/AATCGTACGTACG/primer-reverse/AATCGTACGTCG
NTDPAL/primer-foward/AATCGTACGTACG/primer-reverse/AATCGTACGTCG/stdout
NTDPAL/primer-foward/AATCGTACGTACG/primer-reverse/AATCGTACGTCG/stderr
NTDPAL/primer-foward/ATAGCATCGA
NTDPAL/primer-foward/ATAGCATCGA/primer-reverse
NTDPAL/primer-foward/ATAGCATCGA/primer-reverse/ATAGCATCGAG
NTDPAL/primer-foward/ATAGCATCGA/primer-reverse/ATAGCATCGAG/stdout
NTDPAL/primer-foward/ATAGCATCGA/primer-reverse/ATAGCATCGAG/stderr
NTDPAL/primer-foward/ATAGCATCGA/primer-reverse/AATCGTACGTCG
NTDPAL/primer-foward/ATAGCATCGA/primer-reverse/AATCGTACGTCG/stdout
NTDPAL/primer-foward/ATAGCATCGA/primer-reverse/AATCGTACGTCG/stderr
\end{lstlisting}
The directories are named after the variables and their values.
\section{Control the execution.}
\subsection{Number of simultaneous jobs.}
The number of concurrent jobs is given with --jobs/-j. By default --jobs is the same as the number of CPU cores. --jobs 0 will run as many jobs in parallel as possible.
\begin{lstlisting}
$ ls *.bam | parallel --verbose -j 2 samtools sort {} sorted_{.}
\end{lstlisting}
output:
\begin{lstlisting}
samtools sort ex1a.bam sorted_ex1a
samtools sort ex1.bam sorted_ex1
samtools sort ex1b.bam sorted_ex1b
samtools sort ex1f.bam sorted_ex1f
samtools sort ex1f-rmduppe.bam sorted_ex1f-rmduppe
samtools sort ex1f-rmdupse.bam sorted_ex1f-rmdupse
samtools sort toy.bam sorted_toy
\end{lstlisting}

\section{Interactiveness.}

\begin{lstlisting}
$ ls *.bam | parallel --verbose --interactive  samtools sort {} sorted_{.}
\end{lstlisting}
output:
\begin{lstlisting}
samtools sort ex1a.bam sorted_ex1a ?...y
samtools sort ex1.bam sorted_ex1 ?...n
samtools sort ex1b.bam sorted_ex1b ?...n
samtools sort ex1f.bam sorted_ex1f ?...n
samtools sort ex1f-rmduppe.bam sorted_ex1f-rmduppe ?...y
samtools sort ex1f-rmdupse.bam sorted_ex1f-rmdupse ?...y
samtools sort toy.bam sorted_toy ?...y

$ ls sorted*.bam
sorted_ex1a.bam
sorted_ex1f-rmduppe.bam
sorted_ex1f-rmdupse.bam
sorted_toy.bam
\end{lstlisting}


%%GNU Parallel can be used to put arguments on the command line for an interactive command such as emacs to edit one file at a time:
%%Or give multiple argument in one go to open multiple files:

\subsection{Timing}
Some jobs do heavy I/O when they start. To avoid a thundering herd GNU Parallel can delay starting new jobs. --delay X will make sure there is at least X seconds between each start:
\begin{lstlisting}
ls *.bam | parallel -j 2 --verbose --delay 2.0 samtools sort {} sorted_{.}
\end{lstlisting}
output:
\begin{lstlisting}
samtools sort ex1a.bam sorted_ex1a
samtools sort ex1.bam sorted_ex1
samtools sort ex1b.bam sorted_ex1b
samtools sort ex1f.bam sorted_ex1f
samtools sort ex1f-rmduppe.bam sorted_ex1f-rmduppe
samtools sort ex1f-rmdupse.bam sorted_ex1f-rmdupse
samtools sort toy.bam sorted_toy
\end{lstlisting}
If jobs taking more than a certain amount of time are known to fail, they can be stopped with --timeout:
\begin{lstlisting}
ls ex1f-rmdupse_sorted.bam ~/tmp/BIGBAM/sorted_.bam ex1a_sorted.bam |\
	parallel --timeout 5.0 --verbose samtools index {} 
samtools index ex1a_sorted.bam
samtools index ex1f-rmdupse_sorted.bam
samtools index /home/lindenb/tmp/BIGBAM/sorted_.bam #no bam.bai
\end{lstlisting}

Based on the runtime of completed jobs GNU Parallel can estimate the total runtime:
\begin{lstlisting}
$ ls ex1f-rmdupse_sorted.bam BIGBAM/sorted_.bam ex1a_sorted.bam |\
	parallel --eta --verbose samtools index {} 
\end{lstlisting}
output:
\begin{lstlisting}
samtools index ex1a_sorted.bam
samtools index ex1f-rmdupse_sorted.bam


Computers / CPU cores / Max jobs to run
1:local / 2 / 2

Computer:jobs running/jobs completed/%of started jobs/Average seconds to complete
local:2/0/100%/0.0s samtools index BIGBAM/sorted_.bam
ETA: 28s 0left 19.00avg  local:0/3/100%/19.0s 
\end{lstlisting}


\subsection{Progress}
GNU Parallel can give progress information with --progress:
\begin{lstlisting}
$ ls *_sorted.bam BIGBAM/sorted_.bam  |\
   parallel --progress  samtools index {}
\end{lstlisting}
output:
\begin{lstlisting}
Computers / CPU cores / Max jobs to run
1:local / 2 / 2

Computer:jobs running/jobs completed/%of started jobs/Average seconds to complete
local:0/8/100%/7.2s 
\end{lstlisting}

A logfile of the jobs completed so far can be generated with --joblog:
\begin{lstlisting}
$ ls *_sorted.bam BIGBAM/sorted_.bam  |\
   parallel --joblog log.txt  samtools index {}

$ cat log.txt
Seq	Host	Starttime	Runtime	Send	Receive	Exitval	Signal	Command
1	:	1381155497.299	0.002	0	0	0	0	samtools index ex1a_sorted.bam
2	:	1381155497.301	0.004	0	0	0	0	samtools index ex1b_sorted.bam
3	:	1381155497.306	0.006	0	0	0	0	samtools index ex1f-rmduppe_sorted.bam
4	:	1381155497.313	0.006	0	0	0	0	samtools index ex1f-rmdupse_sorted.bam
5	:	1381155497.319	0.009	0	0	0	0	samtools index ex1f_sorted.bam
6	:	1381155497.326	0.006	0	0	0	0	samtools index ex1_sorted.bam
8	:	1381155497.334	0.004	0	0	0	0	samtools index toy_sorted.bam
7	:	1381155497.332	59.246	0	0	0	0	samtools index BIGBAM/sorted_.bam
\end{lstlisting}
The log contains the job sequence, which host the job was run on, the start time and run time, how much data was transferred if the job was run on a remote host, the exit value, the signal that killed the job, and finally the command being run.


Same command with --timeout
\begin{lstlisting}
$ ls *_sorted.bam BIGBAM/sorted_.bam  |\
   parallel --timeout 5.0 --joblog log.txt  samtools index {}

$ cat log.txt
Seq	Host	Starttime	Runtime	Send	Receive	Exitval	Signal	Command
1	:	1381155638.224	0.002	0	0	0	0	samtools index ex1a_sorted.bam
2	:	1381155638.225	0.008	0	0	0	0	samtools index ex1b_sorted.bam
3	:	1381155638.233	0.006	0	0	0	0	samtools index ex1f-rmduppe_sorted.bam
4	:	1381155638.235	0.009	0	0	0	0	samtools index ex1f-rmdupse_sorted.bam
5	:	1381155638.243	0.007	0	0	0	0	samtools index ex1f_sorted.bam
6	:	1381155638.248	0.008	0	0	0	0	samtools index ex1_sorted.bam
8	:	1381155638.258	0.003	0	0	0	0	samtools index toy_sorted.bam
7	:	1381155638.255	6.188	0	0	-1	15	samtools index BIGBAM/sorted_.bam
\end{lstlisting}
With a joblog GNU Parallel can be stopped and later pickup where it left off. It it important that the input of the completed jobs is unchanged.
\begin{lstlisting}
$ parallel --joblog log.txt  bwa aln -f 0{}_R.fastq.sai toy.fa 0{}_R_.fastq.gz  ::: 1 2
$ cat log.txt
Seq	Host	Starttime	Runtime	Send	Receive	Exitval	Signal	Command
1	:	1381156165.018	0.012	0	0	0	0	bwa aln -f 01_R.fastq.sai toy.fa 01_R_.fastq.gz
2	:	1381156165.028	0.016	0	0	0	0	bwa aln -f 02_R.fastq.sai toy.fa 02_R_.fastq.gz
\end{lstlisting}
With a joblog GNU Parallel can be stopped and later pickup where it left off. It it important that the input of the completed jobs is unchanged.
\begin{lstlisting}
$  parallel --resume --joblog log.txt  \
 bwa aln -f 0{}_R.fastq.sai toy.fa 0{}_R_.fastq.gz  ::: 1 2 3 DOESNTEXIST 4 5
$ cat log.txt

Seq	Host	Starttime	Runtime	Send	Receive	Exitval	Signal	Command
1	:	1381156165.018	0.012	0	0	0	0	bwa aln -f 01_R.fastq.sai toy.fa 01_R_.fastq.gz
2	:	1381156165.028	0.016	0	0	0	0	bwa aln -f 02_R.fastq.sai toy.fa 02_R_.fastq.gz
3	:	1381156215.782	0.014	0	0	0	0	bwa aln -f 03_R.fastq.sai toy.fa 03_R_.fastq.gz
5	:	1381156215.800	0.026	0	0	0	0	bwa aln -f 04_R.fastq.sai toy.fa 04_R_.fastq.gz
6	:	1381156215.830	0.014	0	0	0	0	bwa aln -f 05_R.fastq.sai toy.fa 05_R_.fastq.gz
4	:	1381156215.791	0.139	0	0	0	6	bwa aln -f 0DOESNTEXIST_R.fastq.sai toy.fa 0DOESNTEXIST_R_.fastq.gz
\end{lstlisting}

With --resume-failed GNU Parallel will re-run the jobs that failed:
\begin{lstlisting}
$ parallel --resume-failed --verbose --joblog log.txt  bwa aln -f 0{}_R.fastq.sai toy.fa 0{}_R_.fastq.gz  ::: 1 2 3 DOESNTEXIST 4 5
bwa aln -f 0DOESNTEXIST_R.fastq.sai toy.fa 0DOESNTEXIST_R_.fastq.gz
[bwa_aln] 17bp reads: max_diff = 2
[bwa_aln] 38bp reads: max_diff = 3
[bwa_aln] 64bp reads: max_diff = 4
[bwa_aln] 93bp reads: max_diff = 5
[bwa_aln] 124bp reads: max_diff = 6
[bwa_aln] 157bp reads: max_diff = 7
[bwa_aln] 190bp reads: max_diff = 8
[bwa_aln] 225bp reads: max_diff = 9
[bwa_seq_open] fail to open file '0DOESNTEXIST_R_.fastq.gz'. Abort!

#create the file 0DOESNTEXIST_R_.fastq.gz 
$ cp 01_R_.fastq.gz 0DOESNTEXIST_R_.fastq.gz

$ parallel --resume-failed --verbose --joblog log.txt  bwa aln -f 0{}_R.fastq.sai toy.fa 0{}_R_.fastq.gz  ::: 1 2 3 DOESNTEXIST 4 5
bwa aln -f 0DOESNTEXIST_R.fastq.sai toy.fa 0DOESNTEXIST_R_.fastq.gz
[bwa_aln] 17bp reads: max_diff = 2
[bwa_aln] 38bp reads: max_diff = 3
[bwa_aln] 64bp reads: max_diff = 4
[bwa_aln] 93bp reads: max_diff = 5
[bwa_aln] 124bp reads: max_diff = 6
[bwa_aln] 157bp reads: max_diff = 7
[bwa_aln] 190bp reads: max_diff = 8
[bwa_aln] 225bp reads: max_diff = 9
[bwa_aln_core] calculate SA coordinate... 0.01 sec
[bwa_aln_core] write to the disk... 0.00 sec
[bwa_aln_core] 1000 sequences have been processed.
[main] Version: 0.7.4-r385
[main] CMD: bwa aln -f 0DOESNTEXIST_R.fastq.sai toy.fa 0DOESNTEXIST_R_.fastq.gz
[main] Real time: 0.014 sec; CPU: 0.012 sec
\end{lstlisting}
\subsection{Termination}
todo

\subsection{Limiting the ressources}
todo

\section{Remote execution}



\section{References}
\begin{itemize}
\item{}
\end{itemize}

\end{document}
